### PR TITLE
New: Run webhint via a browser extension

### PR DESCRIPTION
**Pull request by antross**
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- [ ] Added/Updated related documentation.
- [ ] Added/Updated related tests.

## Short description of the change(s)

Refs webhintio/rfcs#48.

Overview of the how the various parts of the extension instantiate and communicate to analyze a website:
sequence

### Testing:
* Download [webhint-browser-extension.zip](https://github.com/webhintio/hint/files/2587392/webhint-browser-extension.zip) (updated 11/15/2018)
* Side-load the zip in your target browser
    * Firefox:
        * Open `about:debugging#addons`
        * Click "Load Temporary Add-on"
        * Choose the zip
    * Edge:
        * Click "..." > "Extensions"
        * Click "Load extension" in the panel that opens
        * Choose the zip
    * Chrome
        * Open `chrome://extensions/`
        * Turn on "Developer mode"
        * Click "Load unpacked"
        * Choose the zip
* Navigate to a target site (e.g. [webhint.io](https://webhint.io/))
* Run webhint
    * Firefox and Chrome
        * Open the developer tools to the new "Hints" tab
        * Click the "Analyze website" button in the tab
    * Edge
        * Open the developer tools to the "Console" tab
        * Click the new webhint icon in the browser toolbar

### Building:
* Checkout the branch for this PR ([`antross/hint:extension-browser`](https://github.com/antross/hint/tree/extension-browser))
* Open a shell at the root of the repo and run
    * `yarn`
    * `yarn build:assets`
    * `cd packages/extension-browser`
    * `yarn webpack`
* Follow the "Testing" steps above to load `packages/extension-browser/dist/bundle` instead of a zip

### Tasks:
* [ ] Validate results match the online scanner where expected (in-progress)
* [ ] Reduce bundle size (tree-shaking + refactoring webhint to use more targeted imports)
* [x] Create a manifest and background script
  * [x] Support starting/stopping
  * [x] Track when content scripts are ready to receive events
  * [x] Capture and forward network traffic to the content script
  * [x] Listen for start/stop requests from devtools page
  * [x] Listen for results from content script and forward to devtools page
  * [x] Forward configuration from devtools page to content script
* [x] Create a connector which can run in a content script
  * [x] Notify background script when ready
  * [x] Forward `fetch::*` events from background-script to `Engine`
  * [x] Generate `traverse::*` events from the page's DOM
  * [x] Support `evaluate` calls from `hint`s
  * [x] Create a static configuration which loads webhint with the new connector
  * [x] Bundle a `webhint.js` content script with the new connector and configuration
  * [x] Forward results to the devtools page via the background script
  * [x] Apply configuration received from devtools page (via background script)
* [x] Add a devtools page to control state and display results
  * [x] Add a tab to the developer tools to display UI
  * [x] Add a button to start/stop and forward request to background script
  * [x] Listen for and display results from background script
  * [x] Add styles to panel and results
  * [x] Add sub-panel for configuration
* [ ] Get relevant hints working
  * [x] `hint-axe`
  * ~`hint-babel-config`~ (omitted as it only applies to local files)
  * [x] `hint-content-type`
  * [x] `hint-disown-opener`
  * [x] `hint-highest-available-document-mode`
  * [x] `hint-http-cache`
  * ~`hint-image-optimization-cloudinary`~ (omitted due to third-party service call)
  * [x] `hint-manifest-app-name`
  * ~`hint-manifest-exists`~ (omitted to avoid assuming PWA by default)
  * [x] `hint-manifest-file-extension`
  * [x] `hint-manifest-is-valid`
  * [x] `hint-meta-charset-utf-8`
  * [x] `hint-meta-viewport`
  * [x] `hint-minified-js`
  * [x] `hint-no-disallowed-headers`
  * [x] `hint-no-html-only-headers`
  * [x] `hint-no-http-redirects`
  * [x] `hint-no-protocol-relative-urls`
  * [ ] `hint-no-vulnerable-javascript-libraries` (needs work to bundle database `json` file)
  * [x] `hint-sri`
  * ~`hint-ssllabs`~ (omitted due to third-party service call)
  * [ ] `hint-strict-transport-security` (has a unique `require` structure that needs adapted)
  * [x] `hint-stylesheet-limits`
  * ~`hint-typescript-config`~ (omitted as it only applies to local files)
  * [x] `hint-validate-set-cookie-header`
  * ~`hint-webpack-config`~ (omitted as it only applies to local files)
  * [x] `hint-x-content-type-options`

### CODE REVIEW

**Comment by antross** [comment]
We now have a basic version (without a devtools page) running in Firefox, Edge, and Chrome:

firefox

edge

chrome

**Comment by antross** [comment]
@digitarald

Just touching base on a few things I'm trying to figure out in Firefox in case you can help.

First, I've run into a known issue where [`devtools.network.onRequestFinished` fires before HAR is populated](https://bugzilla.mozilla.org/show_bug.cgi?id=1472653) in Firefox. This is keeping some hints from functioning correctly in the webhint devtools extension. I may be able to work around it using [`devtools.network.getHAR()`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/devtools.network/getHAR) to get the data later, but thought you'd want to know regardless.

Also I've been looking for a way to link to a location in a resource in the Debugging panel. Chrome has [`devtools.panels.openResource`](https://developer.chrome.com/extensions/devtools_panels#method-openResource) which gets me most of the way there (opens a resource to a line number, but doesn't take a column number). However this API isn't implemented yet in Firefox (as you already know since [you recently filed a bug on it](https://bugzilla.mozilla.org/show_bug.cgi?id=1503475)). Currently I'm falling back to launching a new tab with the `view-source:` protocol. Do you know if there's a better alternative?

Similarly I'd love to link to a resource in the Network panel when webhint has reported issues related to HTTP headers. Thus far I haven't found a way to do this in any browser.

Thanks!

**Comment by janodvarko** [comment]
> First, I've run into a known issue where [`devtools.network.onRequestFinished` fires before HAR is populated](https://bugzilla.mozilla.org/show_bug.cgi?id=1472653) in Firefox.

Do you have a solid test case we could use when fixing the issue. I recall that it's been hard to reproduce that on my machine.

> Currently I'm falling back to launching a new tab with the `view-source:` protocol. Do you know if there's a better alternative?

I don't know about better way to do it atm

> Similarly I'd love to link to a resource in the Network panel when webhint has reported issues related to HTTP headers. Thus far I haven't found a way to do this in any browser.

Correct, this is not supported. I'd be curious how hard it would be to design new WS API and sync between Chrome & Firefox teams. Is there any place where we can suggest new API? Looks like we need better `openResource`

Honza

File: packages/extension-browser/README.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# webhint browser extension (`@hint/extension-browser`)
+
+A web extension and connector to analyze a page running in a browser
+with `webhint`.
```
Comment (molant) [inline packages/extension-browser/README.md]: Maybe add a new issue to have more documentation? We should do some work in the website as well to get this data.

File: packages/extension-browser/package.json [hunk 2]
```diff
@@ -0,0 +1,100 @@
+{
```
Comment (molant) [inline packages/extension-browser/package.json]: Can you ignore the extension when testing for now until we have a solution?

File: packages/extension-browser/src/content-script/connector.ts [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        }
+
+        browser.runtime.onMessage.addListener(async (events: Events) => {
+            if (events.fetchEnd) {
+                await this.notifyFetch(events.fetchEnd);
+            }
+            if (events.fetchStart) {
+                await this._engine.emitAsync('fetch::start', events.fetchStart);
+            }
+            // TODO: Trigger 'fetch::start::target'.
```
Comment (molant) [inline packages/extension-browser/src/content-script/connector.ts]: Maybe the first message is the `target`? What we do in other connectors is to process the messages queued up first and then the rest and assume the first one is the target. This can be part of a separate PR BTW.

File: packages/extension-browser/src/devtools/panel/views/partials/hint.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    const hintStatus = !problems.length ?
+        'PASSED' :
+        `${problems.length} hint${problems.length !== 1 ? 's' : ''}`;
+
+    return html`
+        <details class="hint">
+            <summary class="hint__summary">
+                <span class="hint__name">
+                    ${name}:
```
Comment (molant) [inline packages/extension-browser/src/devtools/panel/views/partials/hint.ts]: You changed this recently, right? Last time I tried it was using the id I think.

**Comment by digitarald** [comment]
Following up on the open question from @molant of how to test the extension in Firefox: We don't have documentation on it but example projects for functional tests using webdriver. Allow me to loop in @rpl, who linked me to https://github.com/mozilla/shield-studies-addon-template/tree/master/test/functional .

Let us know if that helped to set it up or if you run into any issues with using webdriver.

cc @miketaylr who was also looking into this.

**Comment by antross** [comment]
Thanks @digitarald, I'll definitely give that a look and let you know. I recently ran across [webextension-geckodriver](https://github.com/webexts/webextensions-geckodriver) which I was thinking of digging in to, but I wouldn't mind getting closer to using webdriver/geckodriver directly.

@janodvarko 
> Do you have a solid test case we could use when fixing the issue. I recall that it's been hard to reproduce that on my machine.

I currently have one using the entire extension, but let me see if I can prune that down to something a bit smaller. I'll try to reply back with a more self-contained repro early next week.

File: packages/extension-browser/src/devtools/panel/views/partials/hint.ts (hunk 4 above)
Comment (antross) [reply packages/extension-browser/src/devtools/panel/views/partials/hint.ts]: Currently this is still the ID, but is set by the formatter so when the new hint metadata is available there's only one spot that needs to change to cut this over to the name:

In `content-script/formatter.ts`:
```ts
    private buildHintResults(resources: HintResources, problems: Problem[], category: string): HintResults[] {
        return this.getHints(resources, category).map((id) => {
            return {
                helpURL: `https://webhint.io/docs/user-guide/hints/hint-${id}/`,
                id,
                name: id, // TODO: Map to a user-friendly name.
                problems: problems.filter((problem) => {
                    return id === problem.hintId;
                })
            };
        });
    }
```

File: packages/extension-browser/README.md (hunk 1 above)
Comment (antross) [reply packages/extension-browser/README.md]: Filed as #1508.

File: packages/extension-browser/package.json (hunk 2 above)
Comment (antross) [reply packages/extension-browser/package.json]: Done.

**Comment by janodvarko** [comment]
> Here's a fairly minimal extension I created to demonstrate the problem:
> [onrequestfinished-bug.zip](https://github.com/webhintio/hint/files/2596690/onrequestfinished-bug.zip)

I was able to reproduce the issue using that extension.
(I also see that you made a comment in the related bug [report](https://bugzilla.mozilla.org/show_bug.cgi?id=1472653))

Thanks!
Honza
